### PR TITLE
Data-driven registry attachments

### PR DESCRIPTION
**Pull request by Matyrobbrt**
This PR implements an API that can be used to attach arbitrary values to registry entries in a data driven fashion.
Attachments can be used by Neo once #66 is figured out in order to provide data driven APIs for tool modifications, food properties, etc.

Prior art: Quilt Registry Entry Attachments.

### CODE REVIEW

**Comment by Commoble** [comment]
I don't see the point? This is trivial to do without any neoforge apis

**Comment by lukebemish** [comment]
This makes them even simpler, and if I understand it right integrates we'll with tags and whatnot. Beyond that, it can/should be used for forge APIs which is a boon

File: src/main/java/net/minecraftforge/network/PlayMessages.java [hunk 1]
```diff
@@ -323,4 +337,109 @@ public FriendlyByteBuf getAdditionalData()
...
+                pValueWriter.accept(buf, key, value);
+            });
+        }
+
+        public static SyncAttachments<?> decode(FriendlyByteBuf buf)
+        {
+            if (!FMLLoader.getDist().isClient())
+            {
+                throw new UnsupportedOperationException("Cannot deserialize SyncAttachments packet on server!");
+            }
```
Comment (Minecraftschurli) [inline src/main/java/net/minecraftforge/network/PlayMessages.java:378]: Why a dist check and not logical side?

File: src/main/java/net/minecraftforge/network/PlayMessages.java [hunk 2]
```diff
@@ -323,4 +337,109 @@ public FriendlyByteBuf getAdditionalData()
...
+            return new SyncAttachments(
+                    key,
+                    () -> decode((ResourceKey)key, newBuf)
+            );
+        }
+
+        private static <R> Map<AttachmentType<?, R>, Map<Object, ?>> decode(ResourceKey<Registry<R>> registryKey, FriendlyByteBuf buf)
+        {
+            final ClientLevel level = Minecraft.getInstance().level;
+            final Registry<R> registry = level.registryAccess().registryOrThrow(registryKey);
```
Comment (Minecraftschurli) [inline src/main/java/net/minecraftforge/network/PlayMessages.java:394]: Why use the level and not the connection which is safer here?

File: src/main/java/net/minecraftforge/network/PlayMessages.java [hunk 3]
```diff
@@ -323,4 +337,109 @@ public FriendlyByteBuf getAdditionalData()
...
+
+            return map;
+        }
+
+        public static <R> void handle(SyncAttachments<R> msg, Supplier<NetworkEvent.Context> ctx)
+        {
+            ctx.get().enqueueWork(() ->
+            {
+                final ClientLevel level = Minecraft.getInstance().level;
+                final Registry<R> registry = level.registryAccess().registryOrThrow(msg.registry);
```
Comment (Minecraftschurli) [inline src/main/java/net/minecraftforge/network/PlayMessages.java:425]: Here again the connection would be better than using the level.

File: src/main/java/net/minecraftforge/registries/attachment/AttachmentLoader.java [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    private record LoadResult<T>(Map<AttachmentTypeKey<?>, AttachmentLoad<?, T>> results)
+    {
+    }
+
+    private record AttachmentLoad<A, T>(
+            LinkedHashMap<Object, A> attachments
+    )
+    {
+
+    }
```
Comment (Minecraftschurli) [inline src/main/java/net/minecraftforge/registries/attachment/AttachmentLoader.java:255]: Inconsistent formatting.

**Comment by Matyrobbrt** [comment]
Superseded by #519

**Pull request closed without merging**